### PR TITLE
UD: Fix hang during connect in case of CREP retransmission.

### PR DESCRIPTION
**Pull request by yosefe**

### CODE REVIEW

**Comment by yosefe** [comment]
@alex-mikheev @brminich pls review

**Comment by mellanox-github** [comment]
Test PASSed.
See http://bgate.mellanox.com/jenkins/job/gh-ucx-pr/2363/ for details.

**Comment by mellanox-github** [comment]
Test PASSed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/1698/ for details (Mellanox internal link).

File: src/uct/ib/ud/base/ud_ep.c [hunk 1]
```diff
@@ -478,18 +478,25 @@ static void uct_ud_ep_rx_creq(uct_ud_iface_t *iface, uct_ud_neth_t *neth)
...
-     * with ACK_REQ flag */
+    /* Discard duplicate CREP */
+    ooo_type = ucs_frag_list_insert(&ep->rx.ooo_pkts, &skb->u.ooo.elem, neth->psn);
+    if (ooo_type != UCS_FRAG_LIST_INSERT_FAST) {
+        ucs_assertv(ooo_type == UCS_FRAG_LIST_INSERT_DUP, "OOO unsupported");
+        return;
+    }
+
+    ep->dest_ep_id = ctl->conn_rep.src_ep_id;
     ep->rx.ooo_pkts.head_sn = neth->psn;
```
Comment (brminich) [inline src/uct/ib/ud/base/ud_ep.c]: should not this line be removed, because you use  ucs_frag_list_insert before?

Comment (yosefe) [inline src/uct/ib/ud/base/ud_ep.c]: indeed

**Comment by mellanox-github** [comment]
Test PASSed.
See http://bgate.mellanox.com/jenkins/job/gh-ucx-pr/2364/ for details.

**Comment by mellanox-github** [comment]
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/1699/ for details (Mellanox internal link).

**Comment by mellanox-github** [comment]
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/1700/ for details (Mellanox internal link).

**Comment by alex-mikheev** [comment]
we need to add a unit test

**Comment by yosefe** [comment]
@MattBBaker can you post ORNL failure?

**Comment by mellanox-github** [comment]
Test FAILed.
See http://bgate.mellanox.com/jenkins/job/gh-ucx-pr/2367/ for details.

**Comment by mellanox-github** [comment]
Test PASSed.
See http://bgate.mellanox.com/jenkins/job/gh-ucx-pr/2370/ for details.

**Comment by yosefe** [comment]
@alex-mikheev can you pls take a look?

**Comment by mellanox-github** [comment]
Test PASSed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/1704/ for details (Mellanox internal link).

**Comment by alex-mikheev** [comment]
:+1:

**Comment by MattBBaker** [comment]
@yosefe Build error, looks like some badness in the test code with unsigned/signed int comparison. I'll post a fix PR real quick.
```
  CXX    uct/gtest-test_ud.o
cc1plus: warnings being treated as errors
In file included from /var/lib/jenkins/jobs/ornl_pod9/workspace/contrib/../test/gtest/common/test_helpers.h:12,
                 from /var/lib/jenkins/jobs/ornl_pod9/workspace/contrib/../test/gtest/common/test.h:10,
                 from /var/lib/jenkins/jobs/ornl_pod9/workspace/contrib/../test/gtest/uct/uct_test.h:16,
                 from /var/lib/jenkins/jobs/ornl_pod9/workspace/contrib/../test/gtest/uct/test_ud.cc:7:
/var/lib/jenkins/jobs/ornl_pod9/workspace/contrib/../test/gtest/common/gtest.h: In function ‘testing::AssertionResult testing::internal::CmpHelperEQ(const char*, const char*, const T1&, const T2&) [with T1 = int, T2 = unsigned int]’:
/var/lib/jenkins/jobs/ornl_pod9/workspace/contrib/../test/gtest/common/gtest.h:18898:   instantiated from ‘static testing::AssertionResult testing::internal::EqHelper<lhs_is_null_literal>::Compare(const char*, const char*, const T1&, const T2&) [with T1 = int, T2 = uint32_t, bool lhs_is_null_literal = false]’
/var/lib/jenkins/jobs/ornl_pod9/workspace/contrib/../test/gtest/uct/test_ud.cc:637:   instantiated from here
/var/lib/jenkins/jobs/ornl_pod9/workspace/contrib/../test/gtest/common/gtest.h:18861: error: comparison between signed and unsigned integer expressions
  CXX    uct/gtest-test_ud_slow_timer.o
make[2]: *** [uct/gtest-test_ud.o] Error 1
make[2]: *** Waiting for unfinished jobs....
make[2]: Leaving directory `/var/lib/jenkins/jobs/ornl_pod9/workspace/build-tests/test/gtest'
make[1]: *** [all-recursive] Error 1
make[1]: Leaving directory `/var/lib/jenkins/jobs/ornl_pod9/workspace/build-tests'
make: *** [all] Error 2
Build step 'Execute shell' marked build as failure
```

**Pull request merged**